### PR TITLE
EZEE-3465: Refactored RepositoryConfigurationProvider

### PR DESCRIPTION
**Pull request by barw4**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZEE-3465](https://jira.ez.no/browse/EZEE-3465)
| **Improvement**| yes
| **New feature**    | no
| **Target version** | `7.5`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

Minor tweaks have been made to allow pulling out the default repository and list of repositories globally.

Related `ezplatform-admin-ui` PR: https://github.com/ezsystems/ezplatform-admin-ui/pull/1704

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Bundle/EzPublishCoreBundle/ApiLoader/RepositoryConfigurationProvider.php [hunk 1]
```diff
@@ -54,6 +51,13 @@ public function getRepositoryConfig()
         return ['alias' => $repositoryAlias] + $this->repositories[$repositoryAlias];
     }
 
+    public function pullDefaultRepository(): ?string
```
Comment (Steveb-p) [inline eZ/Bundle/EzPublishCoreBundle/ApiLoader/RepositoryConfigurationProvider.php:54]: Shouldn't this be `private`?

File: eZ/Bundle/EzPublishCoreBundle/ApiLoader/RepositoryConfigurationProvider.php [hunk 2]
```diff
@@ -54,6 +51,13 @@ public function getRepositoryConfig()
...
+    {
+        $aliases = array_keys($this->repositories);
+
+        return array_shift($aliases);
```
Comment (adamwojs) [inline eZ/Bundle/EzPublishCoreBundle/ApiLoader/RepositoryConfigurationProvider.php:63]: Could be replaced by https://www.php.net/manual/en/function.array-key-first during merge up to 3.x

File: eZ/Bundle/EzPublishCoreBundle/ApiLoader/RepositoryConfigurationProvider.php [hunk 3]
```diff
@@ -62,4 +66,9 @@ public function getStorageConnectionName(): string
             ? $repositoryConfig[self::REPOSITORY_STORAGE][self::REPOSITORY_CONNECTION]
             : self::DEFAULT_CONNECTION_NAME;
     }
+
+    public function getRepositories(): array
```
Comment (Steveb-p) [inline eZ/Bundle/EzPublishCoreBundle/ApiLoader/RepositoryConfigurationProvider.php]: Please provide a phpdoc with returned type.

File: eZ/Bundle/EzPublishCoreBundle/ApiLoader/RepositoryConfigurationProvider.php (hunk 1 above)
Comment (adamwojs) [inline eZ/Bundle/EzPublishCoreBundle/ApiLoader/RepositoryConfigurationProvider.php]: Could you please rename it to `getDefaultRepositoryAlias()` ?    
  
Method `getCurrentRepositoryAlias` covering this logic [https://github.com/ezsystems/ezplatform-admin-ui/pull/1704/files#diff-e54466cbf9baa7083eef49e3874011f88bc048e74724640edfa77685dab2c527R55](https://github.com/ezsystems/ezplatform-admin-ui/pull/1704/files#diff-e54466cbf9baa7083eef49e3874011f88bc048e74724640edfa77685dab2c527R55) would be very useful in other contexts :wink:

**Review by adamwojs — CHANGES_REQUESTED** [review]

**Review by alongosz — CHANGES_REQUESTED** [review]
Besides Adam's remarks: missing unit test coverage.

File: eZ/Bundle/EzPublishCoreBundle/Tests/ApiLoader/RepositoryConfigurationProviderTest.php [hunk 4]
```diff
@@ -93,6 +93,26 @@ public function testGetRepositoryConfigUndefinedRepository()
...
+            'second' => [
+                'engine' => 'bar',
+            ],
+        ];
+
+        $provider = new RepositoryConfigurationProvider($configResolver, $repositories);
+        $provider->getRepositoryConfig();
+
+        $this->assertSame('first', $provider->getDefaultRepositoryAlias());
+        $this->assertSame('first', $provider->getCurrentRepositoryAlias());
```
Comment (adamwojs) [inline eZ/Bundle/EzPublishCoreBundle/Tests/ApiLoader/RepositoryConfigurationProviderTest.php]: Well, this is unit tests which means that each code unit should be tested separately. I recommend to split getDefaultRepositoryAlias() and getCurrentRepositoryAlias() tests.

**Review by adamwojs — APPROVED** [review]

File: eZ/Bundle/EzPublishCoreBundle/Tests/ApiLoader/RepositoryConfigurationProviderTest.php (hunk 4 above)
Comment (barw4) [reply eZ/Bundle/EzPublishCoreBundle/Tests/ApiLoader/RepositoryConfigurationProviderTest.php]: Resolved in: https://github.com/ezsystems/ezpublish-kernel/pull/3085/commits/c6f4d0939bcaec353c790d40a8dd30cf9140cc22

File: eZ/Bundle/EzPublishCoreBundle/Tests/ApiLoader/RepositoryConfigurationProviderTest.php [hunk 5]
```diff
@@ -93,6 +87,48 @@ public function testGetRepositoryConfigUndefinedRepository()
...
+     * @dataProvider providerForRepositories
+     */
+    public function testGetDefaultRepositoryAlias(array $repositories): void
+    {
+        $configResolver = $this->getConfigResolverMock();
+
+        $provider = new RepositoryConfigurationProvider($configResolver, $repositories);
+        $provider->getRepositoryConfig();
+
+        $this->assertSame('first', $provider->getDefaultRepositoryAlias());
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Tests/ApiLoader/RepositoryConfigurationProviderTest.php]: EA: use self for static methods:
```suggestion
        self::assertSame('first', $provider->getDefaultRepositoryAlias());
```

File: eZ/Bundle/EzPublishCoreBundle/Tests/ApiLoader/RepositoryConfigurationProviderTest.php [hunk 6]
```diff
@@ -93,6 +87,48 @@ public function testGetRepositoryConfigUndefinedRepository()
...
+     * @dataProvider providerForRepositories
+     */
+    public function testGetCurrentRepositoryAlias(array $repositories): void
+    {
+        $configResolver = $this->getConfigResolverMock();
+
+        $provider = new RepositoryConfigurationProvider($configResolver, $repositories);
+        $provider->getRepositoryConfig();
+
+        $this->assertSame('first', $provider->getCurrentRepositoryAlias());
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Tests/ApiLoader/RepositoryConfigurationProviderTest.php]:
```suggestion
        self::assertSame('first', $provider->getCurrentRepositoryAlias());
```

**Review by alongosz — CHANGES_REQUESTED** [review]

**Review by micszo — APPROVED** [review]
QA Approved on eZ Platform EE v2.5.16 with diffs.

**Comment by adamwojs** [comment]
@barw4 Could you please merge up changes?

**Comment by barw4** [comment]
Merged into: 
`1.2` => https://github.com/ezsystems/ezplatform-kernel/commit/139715b22d5ade5fd6a74eddfbbbfee313f22802
`master` => https://github.com/ezsystems/ezplatform-kernel/commit/19f134e070285b0154b5f7d300cf360522c71aa1

**Pull request merged**